documentation/kernel-semantics.tex

```diff
@@ -541,18 +542,52 @@ \subsubsection{Variable Lookup}
 \subsection{Main Environment}
 \label{subsec:main-env}
 
-The main environment denoted as $\mainenv$ is a function that maps static and library members to locations.
-\[
-    \mainenv \in \dmenv = \dfield \rightarrow \dlocation.
-\]
-We introduce the main environment to support static and library field extraction and assignment.
+\kernel{} supports static and library fields, which are initialized when first accessed and mutations of values stored in these fields are visible in all subsequent execution of statements or evaluation of expressions.
+To support this feature, we introduce a main environment, denoted as $\mainenv$.
+
+\newcommand{\dmainenv}{\textbf{MainEnvironment}}
+
+\[\mainenv \in \dmainenv = \dmember \rightarrow \dlocation \]
+
+We define the following functions for manipulating $\mainenv$:
+
+\subsubsection{Contains}
+We define the function $\contains$ as follows:
+\begin{align*}
+    \contains &\in \dmainenv \times \dmember \rightarrow \dsembool\\
+    \contains(\mainenv, x) &=\begin{cases}
+        \true &\text{ if }x \in \textit{dom}(\mainenv)\\
+        \false &\text{ otherwise}
+    \end{cases}
+\end{align*}
+where $\textit{dom}(\mainenv)$ is the domain of the environment $\mainenv$, i.e. the set of members for which $\mainenv$ is defined.
+
+\subsubsection{Lookup}
+The function $\lookup$ is defined as follows:
+\begin{align*}
+    \lookup &\in \dmainenv \times \dmember \rightarrow \dlocation\\
+    \lookup(\mainenv, x) &= \mainenv(x)\\
+\end{align*}
+where $x \in \textit{dom}(\mainenv)$.
+Note that the function $\lookup$ is defined only for elements of $\textit{dom}(\mainenv)$.
+
+\subsubsection{Extend}
+We define the function extend as follows:
+\begin{align*}
+    &\extend \in \dmainenv \times \dmember \times \dval \rightarrow \denv\\
+    &\extend(\mainenv, x, \val) = \mainenv'\\
+    &\text{where }\\
+    &\hspace{2cm}\forall y \in \textit{dom}(\mainenv) \cup x,\hspace{2mm}\mainenv'(y) =\begin{cases}
```

Comment:
Please avoid using commands for exact formatting like `\hspace{2cm}` until the text is mostly ready, i.e. until formatting is the only task left for the spec.